core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -615,4 +630,49 @@ private static void updateTotal(
       }
     }
   }
+
+  private Long snapshotIdToRollbackToOnConflict(Long parentSnapshotId) {
+    Long newParentSnapshotId = parentSnapshotId;
+    if (shouldRollbackReplaceOnConflict()) {
+      // add a set snapshot op on top of base to roll back to parent snapshot
+      boolean isCommitSuccessfullyApplied = false;
+      // Update parentSnapshot to it's grandParent
+      // provided parentSnapshot is of type replace and never rollback beyond startingSnapshotId
+      while (newParentSnapshotId != null
+          && DataOperations.REPLACE.equals(base.snapshot(newParentSnapshotId).operation())
+          && !newParentSnapshotId.equals(startingSnapshotId())) {
+        // create a tempTableOperation to pass base with rollback to validate of update
+        TableOperations tempTableOps = ops.temp(base);
+        newParentSnapshotId = base.snapshot(newParentSnapshotId).parentId();
+        if (newParentSnapshotId == null) {
+          return null;
+        }
+        Snapshot parentSnapshot = base.snapshot(newParentSnapshotId);
+
+        SetSnapshotOperation setSnapshotOp = new SetSnapshotOperation(tempTableOps);
+        setSnapshotOp.rollbackTo(newParentSnapshotId).commit();
+        try {
+          validate(tempTableOps.current(), parentSnapshot);
+          isCommitSuccessfullyApplied = true;
+        } catch (ValidationException validationException) {
+          // swallow the exception for re-trying
+        }
+      }
+
+      if (!isCommitSuccessfullyApplied) {
+        return null;
+      }
+
+      return newParentSnapshotId;
+    } else {
+      return null;
+    }
+  }
+
+  private Boolean shouldRollbackReplaceOnConflict() {
+    return PropertyUtil.propertyAsBoolean(
+        base.properties(),
+        TableProperties.COMMIT_ROLLBACK_REPLACE_ON_CONFLICT_ENABLED,
```

Comment:
I am a bit not sure if this should be a table level property. I would imagine this to enabled only on certain sessions, but the majority of writers still do not rollback. Is there any other option to make this a runtime config instead?

Reply:
Explored a bit more offline, I think maybe it's better to do this through an update in related action APIs, for example:

```
public interface SupportsRollbackReplace<T> {
  
  T rollbackReplaceOnConflict();
}

public interface AppendFiles extends SnapshotUpdate<AppendFiles>, SupportsRollbackReplace<AppendFiles> {
  ...
}
```

So that in core we can do:

```
table.newAppend().rollbackReplaceOnConflict().addFiles(...).commit()
```

And then we can make this a session config or runtime option in engine.

What do you think?

Reply:
we could potentially make this more generic, for example just add this to `SnapshotUpdate<T>`

```
  T rollback(List<String> operations);
```

Although I don't know what would be the use case of the other kinds of rollback, but seems like a good generalization.

@rdblue do you have any thoughts and further comments regarding this PR?